auction/program/src/processor.rs

```diff
@@ -0,0 +1,299 @@
+    max_winners: usize,
+    /// Pubkey of the resource being bid on.
+    resource: Pubkey,
+    /// Time the auction starts at, this may be changed only if the auction hasn't started.
+    start_time: UnixTimestamp,
+}
+
+/* -------------------------------------------------------------------------------- */
+
+/// Arguments for the CreateAuction instruction discriminant.
+#[repr(C)]
+#[derive(Clone, BorshSerialize, BorshDeserialize, PartialEq)]
+pub struct CreateAuctionArgs {
+    /// The start time requested for this auction. See AuctionData.
+    start_time: UnixTimestamp,
+    /// How many winners are allowed for this auction. See AuctionData.
+    max_winners: usize,
+    /// The resource being auctioned. See AuctionData.
+    resource: Pubkey,
+}
+
+/// Arguments for the PlaceBid instruction discriminant .
+#[repr(C)]
+#[derive(Clone, BorshSerialize, BorshDeserialize, PartialEq)]
+pub struct PlaceBidArgs {
+    /// Size of the bid being placed. The user must have enough SOL to satisfy this amount.
+    pub amount: u64,
+}
+
+#[repr(C)]
+#[derive(Clone, BorshSerialize, BorshDeserialize, PartialEq)]
+pub struct CancelBidArgs {}
+
+#[repr(C)]
+#[derive(Clone, BorshSerialize, BorshDeserialize, PartialEq)]
+pub enum AuctionInstruction {
+    CreateAuction(CreateAuctionArgs),
+    PlaceBid(PlaceBidArgs),
+    CancelBid(CancelBidArgs),
+}
+
+pub fn process_instruction(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    input: &[u8],
+) -> ProgramResult {
+    let account_iter = &mut accounts.iter();
+
+    match AuctionInstruction::try_from_slice(input)? {
+        AuctionInstruction::CreateAuction(args) => create_auction(program_id, accounts, args),
+        AuctionInstruction::PlaceBid(args) => place_bid(program_id, accounts, args),
+        AuctionInstruction::CancelBid(args) => cancel_bid(program_id, accounts),
+    };
+
+    Ok(())
+}
+
+/// Creates a new auction account. This will verify the start time is valid, and that the resource
+/// being bid on exists. The creator of the auction by default has authority to modify the auction
+/// state, including setting someone else as the auction authority.
+///
+/// Possible methods to store bid data.
+///
+/// 1) Store the entire bid history in the auction account itself with a list.
+/// 2) Use a counter for total number of bids, and use PDAs to store individual bids.
+/// 3) Create a ring buffer the size of the winner list, and throw away cancelled bids.
+///
+/// For now going with 1 for ease of implementation, will come back to this to figure out cost
+/// and/or efficiency of the optoins.
+fn create_auction(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    args: CreateAuctionArgs,
+) -> ProgramResult {
+    let account_iter = &mut accounts.iter();
+    let creator_act = next_account_info(account_iter)?;
+    let auction_act = next_account_info(account_iter)?;
+
+    let auction_path = [
+        PREFIX.as_bytes(),
+        program_id.as_ref(),
```

Comment:
With PDA-based accounts, the front end won't have the ability to allocate to them, since it won't have authority, you'll need to do the allocating.